xep-0407.xml

```diff
@@ -172,20 +174,24 @@
       <p>
-        A MIX channel MAY support message retraction, where the sender of a messages or an authorized administrator deletes a message.  A MIX channel MAY limit the time frame in which a message is allowed to be retracted, for example to prevent retraction of very old messages.   When a messages is retracted the original message MAY be replaced by a tombstone.  Message retraction is done by sending a special message that identifies the original message.   This mechanism allows the retraction to be distributed on the same path as the original message so that all participating servers and clients MAY honour the retraction.  The protocol to request retraction does this by adding to a message  a &lt;retract&gt; element qualified by the 'urn:xmpp:mix:misc:0' namespace.  A retract messages MUST NOT have a body.   The &lt;retract&gt; element MUST include an 'id' attribute that holds the MAM-ID of the original  message.  The message sender will need to look up the MAM-ID.   The MAM-ID is the convenient message identification for message recipients.  A message and its retraction shown in the following example.
+        A MIX channel MAY support message retraction, where the sender of a messages or an authorized administrator deletes a message.  A MIX channel MAY limit the time frame in which a message is allowed to be retracted, for example to prevent retraction of very old messages.   When a messages is retracted the original message MAY be replaced by a tombstone.  Message retraction is done by sending a special message that identifies the original message.   This mechanism allows the retraction to be distributed on the same path as the original message so that all participating servers and clients MAY honour the retraction.  
+        
+        The protocol to request retraction is defined in &xep0424;. If a server is aware of message retractions it MUST advertise it's support by advertising  'urn:xmpp:message-retract:0' feature. The retract message MUST be addressed to the mix channel and the &lt;apply-to&gt; element (as defined in &xep0422;) MUST include an 'id' attribute that holds the stable-id (as defined in &xep0359;) of the original message send by the channel itself. A message and its retraction shown in the following example.
       </p>
       <example caption="User Retracts a Message"><![CDATA[
-
 <message from='hag66@shakespeare.example/UUID-a1j/7533'
          to='coven@mix.shakespeare.example'
          id='abcde'>
   <body> A Message I did not mean to send </body>
+  <stanza-id by='hag66@shakespeare.example' id='a568706e-867b-49f1-a31b-e428ba9cece1' xmlns='urn:xmpp:sid:0'/>
 </message>
 
 <message from='hag66@shakespeare.example/UUID-a1j/7533'
          to='coven@mix.shakespeare.example'
          id='92vax143g'>
-  <retract id='77E07BB0-55CF-4BD4-890E-3F7C0E686BBD' xmlns='urn:xmpp:mix:misc:0'/>
+  <apply-to id="a568706e-867b-49f1-a31b-e428ba9cece1" xmlns="urn:xmpp:fasten:0">
```

Comment:
I think this could be solved if apply-to would not only refer to the id, but also to the 'by' value of the stanza-id element. For example
```
<apply-to
  xmlns="urn:xmpp:fasten:0"
  id="a568706e-867b-49f1-a31b-e428ba9cece1"
  by='hag66@shakespeare.example'>
```

See also my 2019 mail to @standards: https://mail.jabber.org/pipermail/standards/2019-September/036435.html

Reply:
I don't think stamping the origin of the id actually matters very much in this case (it very much does in others), because in this case only one id is going to be usable - the one from the MIX itself, so the stamper can be implicit.

Reply:
Maybe, but I wonder how many implementations will get the implicit value wrong and open a security hole due that. I am usually a fan of avoiding unnecessary bytes on the wire and implicit values. But this smells a little bit like the carbon 'from' security hole that many implementations had.

I think at least, it should be clearly spelled out that the message is fastened to a message containing a stanza-id element where the 'id' matches the 'id' value from apply-to, and the 'by' value is the bare JID of the MIX channel (right?).

Reply:
I'm with @Kev here - the example was just bad and used wrong ID. Adding "by" wouldn't change that here.